cmd/syncthing/gui.go

```diff
@@ -386,32 +396,26 @@ func (s *apiService) VerifyConfiguration(from, to config.Configuration) error {
 }
 
 func (s *apiService) CommitConfiguration(from, to config.Configuration) bool {
-	if to.GUI == from.GUI {
+	if to.GUI.Address() == from.GUI.Address() {
 		return true
 	}
 
-	// Order here is important. We must close the listener to stop Serve(). We
-	// must create a new listener before Serve() starts again. We can't create
-	// a new listener on the same port before the previous listener is closed.
-	// To assist in this little dance the Serve() method will wait for a
-	// signal on the configChanged channel after the listener has closed.
+	// This is a bit tricky. We can't test the new configuration without creating
+	// a new listener. We can't create a new listener without stopping the first
+	// listener. If the new listener creation fails, then because the GUI no
+	// longer works we'll never tell the user.
+	// There also aren't any errors in creating a new listener which restarting
+	// will fix. The only bit of config we use is the listen address, and either
+	// this is invalid in some way or something's already listening on that port.
 
-	s.listenerMut.Lock()
-	defer s.listenerMut.Unlock()
-
-	s.listener.Close()
+	// Therefore there's no point in testing whether a new listener can be created
+	// at this stage.
 
-	var err error
-	s.listener, err = s.getListener(to.GUI)
-	if err != nil {
-		// Ideally this should be a verification error, but we check it by
-		// creating a new listener which requires shutting down the previous
-		// one first, which is too destructive for the VerifyConfiguration
-		// method.
-		return false
-	}
+	s.listenAddressChanged <- to.GUI.Address()
 
-	s.configChanged <- struct{}{}
+	s.listenerMut.Lock()
+	s.listener.Close()
+	s.listenerMut.Unlock()
```

Comment:
It won't pop up in practice (I think?) as T2 is blocked on `listener.Serve()` until T1 calls `s.listener.Close()`. 

The same pattern is used in `Stop()`:

```go
func (s *apiService) Stop() {
	s.listenerMut.Lock()
	listener := s.listener
	s.listenerMut.Unlock()

	close(s.stop)

	// listener may be nil here if we've had a config change to a broken
	// configuration, in which case we shouldn't try to close it.
	if listener != nil {
		listener.Close()
	}
}
```

Or am I misunderstanding you?